lock/session_locker.go

```diff
@@ -0,0 +1,23 @@
+// Package lock defines the Locker interface and implements the locking logic.
+package lock
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+)
+
+var (
+	// ErrLockNotImplemented is returned when the database does not support locking.
+	ErrLockNotImplemented = errors.New("lock not implemented")
+	// ErrUnlockNotImplemented is returned when the database does not support unlocking.
+	ErrUnlockNotImplemented = errors.New("unlock not implemented")
+)
+
+// SessionLocker is the interface to lock and unlock the database for the duration of a session. The
+// session is defined as the duration of a single connection and both methods must be called on the
+// same connection.
+type SessionLocker interface {
```

Comment:
I took the best bits of https://github.com/pressly/goose/pull/575 and ported them into this PR.

Kudos to @roblaszczak for suggesting a common interface so users can bring their own implementation. I think this should suffice? Feedback/suggestions are welcome.

I called this "SessionLocker" with the hopes of one day also adding a "TransactionLocker".

The one gotcha here is supplying a locker that doesn't match up with the correct dialect. I tried to name the functions appropriately, but there's nothing enforcing this right now.